### PR TITLE
Refactoring of Jump and Drop models

### PR DESCRIPTION
**Pull request by VargBurz**
This PR adds parent `Stair` class for `Jump` and `Drop` models.

### Motivation:
`Jump` and `Drop` models are almost identical. We need to remove code duplication which presents because `Jump` algorithm is a reverted `Drop` algorithm.

## Changes:
 - new  `StairModel` class
 - `models/jump_model.py`, `models/drop_model.py`:
      - inherit `Jump`/`Drop` classes  from `StairModel` class
      - `Jump`/`Drop` methods `do_fit`, `do_detect`, `filter_detection`  are `StairModel` class methods now
      - new `get_extremum_type` method
 - `models/model.py`:
      - new `ModelType` enum: `jump`, `drop`, `peak`, `trough`, `general`
      - new `ExtremumType` enum: `min` or `max`, used to connect models with utils algorithms
      - fix imports order
      - `get_model_type` method returns `ModelType`
      - use new `remove_duplicates_and_sort` utils method instead of `list(set(array[]))` 
      - `_update_fiting_result` -> `_update_fitting_result`
 - `utils/common.py `: 
      - move `find_drop` and `find_jump` methods to `Stair` model, rename to `get_stair_indexes`
      - new `remove_duplicates_and_sort` method
      - `get_parameters_from_segments`: remove unused `model_type` arg
 - `tests/test_models.py`:  tests for `get_stair_indexes` method
 - `tests/test_utils.py`: test for `remove_duplicates_and_sort` method

### CODE REVIEW

File: analytics/analytics/models/drop_model.py [hunk 1]
```diff
@@ -41,82 +42,14 @@ def find_segment_center(self, dataframe: pd.DataFrame, start: int, end: int) ->
...
+        '''
+        Find drop indexes
+        '''
+        #TODO: refactor and move method to stair_model
+        indexes = []
+        for i in range(len(data)-length-1):
+            for x in range(1, length):
+                if(data[i + x] < data[i] - height):
+                    indexes.append(i)
+        return(indexes)
```
Comment (rozetko) [inline analytics/analytics/models/drop_model.py]: It's not js, lol
```suggestion
        return indexes
```

File: analytics/analytics/models/drop_model.py [hunk 2]
```diff
@@ -41,82 +42,14 @@ def find_segment_center(self, dataframe: pd.DataFrame, start: int, end: int) ->
...
-        for item in delete_list:
-            segments.remove(item)
-        return set(segments)
+    def get_stair_indexes(self, data: pd.Series, height: float, length: int) -> List[int]:
+        '''
+        Find drop indexes
+        '''
+        #TODO: refactor and move method to stair_model
+        indexes = []
+        for i in range(len(data)-length-1):
```
Comment (rozetko) [inline analytics/analytics/models/drop_model.py]:
```suggestion
        for i in range(len(data) - length - 1):
```

File: analytics/analytics/models/drop_model.py [hunk 3]
```diff
@@ -41,82 +42,14 @@ def find_segment_center(self, dataframe: pd.DataFrame, start: int, end: int) ->
...
-                        delete_list.append(segment)
-                except ValueError:
-                    delete_list.append(segment)
-            else:
-                delete_list.append(segment)
-
-        for item in delete_list:
-            segments.remove(item)
-        return set(segments)
+    def get_stair_indexes(self, data: pd.Series, height: float, length: int) -> List[int]:
```
Comment (rozetko) [inline analytics/analytics/models/drop_model.py]: please add a docstring

File: analytics/tests/test_models.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import pandas as pd
+import numpy as np 
+import models
+
+class TestModel(unittest.TestCase):
+
+    def test_stair_model_get_indexes(self):
+        drop_model = models.DropModel()
+        jump_model = models.JumpModel()
+        drop_data = pd.Series([4,4,4,1,1,1])
```
Comment (rozetko) [inline analytics/tests/test_models.py]: let's also add a case with more than 1 drop / jump

File: analytics/analytics/models/stair_model.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                    continue
+                elif 0 < percent_of_nans <= 0.5:
+                    nan_list = utils.find_nan_indexes(convol_data)
+                    convol_data = utils.nan_to_zero(convol_data, nan_list)
+                    pattern_data = utils.nan_to_zero(pattern_data, nan_list)
+                conv = scipy.signal.fftconvolve(convol_data, pattern_data)
+                upper_bound = self.state.convolve_max * 1.2
+                lower_bound = self.state.convolve_min * 0.8
+                delete_up_bound = self.state.conv_del_max * 1.02
+                delete_low_bound = self.state.conv_del_min * 0.98
```
Comment (rozetko) [inline analytics/analytics/models/stair_model.py]: I'm wondering what are 1.2, 0.8, 1.02, 0.98 :)

File: analytics/analytics/models/model.py [hunk 6]
```diff
@@ -6,12 +6,20 @@
...
 import logging
+from enum import Enum
+
 from analytic_types import AnalyticUnitId, ModelCache, TimeSeries
 from analytic_types.segment import Segment
 from analytic_types.learning_info import LearningInfo
 
 import utils.meta
 
+class ModelName(Enum):
```
Comment (rozetko) [inline analytics/analytics/models/model.py]: do you use it anywhere?

File: analytics/analytics/models/stair_model.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        '''
+        data: data, that contains stair (jump or drop) segments,
+        length: the number of indexes to be contained in the stair segment,
+        height: the difference between stair max_line and min_line(see utils.find_parameters),
+        return: list of start stair segment indexes
+        '''
+        #TODO: refactor and move method to stair_model
+        indexes = []
+        comparison_operator = operator.gt
+        if self.get_model_type()[0] == 'drop':
```
Comment (rozetko) [inline analytics/analytics/models/stair_model.py]: enum?

File: analytics/analytics/models/stair_model.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        del_conv_list = []
+        delete_pattern_timestamp = []
+        for segment in deleted_segments:
+            segment_cent_index = segment.center_index
+            delete_pattern_timestamp.append(segment.pattern_timestamp)
+            deleted_stair = utils.get_interval(data, segment_cent_index, window_size)
+            deleted_stair = utils.subtract_min_without_nan(deleted_stair)
+            del_conv_stair = scipy.signal.fftconvolve(deleted_stair, self.state.pattern_model)
+            if len(del_conv_stair): del_conv_list.append(max(del_conv_stair))
```
Comment (jonyrock) [inline analytics/analytics/models/stair_model.py]: 1) if `len(del_conv_stair)` what? please be explicit
2) please add a new line, this is our codestyle

File: analytics/analytics/models/stair_model.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                upper_bound = self.state.convolve_max * (1 + POSITIVE_SEGMENT_MEASUREMENT_ERROR)
+                lower_bound = self.state.convolve_min * (1 - POSITIVE_SEGMENT_MEASUREMENT_ERROR)
+                delete_up_bound = self.state.conv_del_max * (1 + NEGATIVE_SEGMENT_MEASUREMENT_ERROR)
+                delete_low_bound = self.state.conv_del_min * (1 - NEGATIVE_SEGMENT_MEASUREMENT_ERROR)
+                try:
+                    if max(conv) > upper_bound or max(conv) < lower_bound:
+                        delete_list.append(segment)
+                    elif max(conv) < delete_up_bound and max(conv) > delete_low_bound:
+                        delete_list.append(segment)
+                except ValueError:
```
Comment (jonyrock) [inline analytics/analytics/models/stair_model.py]: remote this logic on exceptions. 
For now, explain when you get `ValueError`

File: analytics/analytics/models/stair_model.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+    def do_fit(
+        self,
+        dataframe: pd.DataFrame,
+        labeled_segments: List[AnalyticSegment],
+        deleted_segments: List[AnalyticSegment],
+        learning_info: LearningInfo
+    ) -> None:
+        data = utils.cut_dataframe(dataframe)
+        data = data['value']
```
[the remark this answers is not in the document]
Comment (rozetko) [reply analytics/analytics/models/stair_model.py:66]: @jonyrock I'd not do it in this PR, it's pretty huge already
Let's do it in a separate PR?

File: analytics/analytics/models/drop_model.py [hunk 11]
```diff
@@ -13,110 +13,29 @@
...
+class DropModel(StairModel):
 
-    def get_model_type(self) -> (str, bool):
-        model = 'drop'
-        type_model = False
-        return (model, type_model)
+    def get_model_type(self) -> ModelName:
+        return ModelName.DROP
 
     def find_segment_center(self, dataframe: pd.DataFrame, start: int, end: int) -> int:
```
Comment (rozetko) [inline analytics/analytics/models/drop_model.py:36]: todo

File: analytics/analytics/models/stair_model.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+    def get_state(self, cache: Optional[dict] = None) -> StairModelState:
+        return StairModelState.from_json(cache)
+
+    def get_stair_indexes(self, data: pd.Series, height: float, length: int) -> List[int]:
+        """Get list of start stair segment indexes.
+
+        Keyword arguments:
+        data -- data, that contains stair (jump or drop) segments
+        length -- the maximum count of values to be contained in the stair
```
Comment (rozetko) [inline analytics/analytics/models/stair_model.py:45]:
```suggestion
        length -- maximum count of values in the stair
```

**Comment by jonyrock** [comment]
what about "ExtremumType"

**Pull request merged**